libfreefare/freefare.3
.\" Copyright (C) 2010 Romain Tartiere
.\"
.\" This program is free software: you can redistribute it and/or modify it
.\" under the terms of the GNU Lesser General Public License as published by the
.\" Free Software Foundation, either version 3 of the License, or (at your
.\" option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful, but WITHOUT
.\" ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
.\" FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
.\" more details.
.\"
.\" You should have received a copy of the GNU Lesser General Public License
.\" along with this program.  If not, see <http://www.gnu.org/licenses/>
.\"
.\" $Id$
.\"
.Dd March 30, 2010
.Dt FREEFARE 3
.Os
.\"  _   _
.\" | \ | | __ _ _ __ ___   ___
.\" |  \| |/ _` | '_ ` _ \ / _ \
.\" | |\  | (_| | | | | | |  __/
.\" |_| \_|\__,_|_| |_| |_|\___|
.\"
.Sh NAME
.Nm freefare_get_tags ,
.Nm freefare_get_tag_type ,
.Nm freefare_get_tag_friendly_name ,
.Nm freefare_get_tag_uid ,
.Nm freefare_free_tag ,
.Nm freefare_free_tags
.Nd Generic target manipulation functions
.\"  _     _ _
.\" | |   (_) |__  _ __ __ _ _ __ _   _
.\" | |   | | '_ \| '__/ _` | '__| | | |
.\" | |___| | |_) | | | (_| | |  | |_| |
.\" |_____|_|_.__/|_|  \__,_|_|   \__, |
.\"                               |___/
.Sh LIBRARY
Mifare card manipulation library (libfreefare, \-lfreefare)
.\"  ____                              _
.\" / ___| _   _ _ __   ___  _ __  ___(_)___
.\" \___ \| | | | '_ \ / _ \| '_ \/ __| / __|
.\"  ___) | |_| | | | | (_) | |_) \__ \ \__ \
.\" |____/ \__, |_| |_|\___/| .__/|___/_|___/
.\"        |___/            |_|
.Sh SYNOPSIS
.In freefare.h
.Ft "MifareTag *"
.Fn freefare_get_tags "nfc_device_t *device"
.Bd -literal
enum mifare_tag_type {
    ULTRALIGHT,
    CLASSIC_1K,
    CLASSIC_4K
};
.Ed
.Ft "enum mifare_tag_type"
.Fn freefare_get_tag_type "MifareTag tag"
.Ft "const char *"
.Fn freefare_get_tag_friendly_name "MifareTag tag"
.Ft "char *"
.Fn freefare_get_tag_uid "MifareTag tag"
.Ft "void"
.Fn freefare_free_tag "MifareTag tags"
.Ft "void"
.Fn freefare_free_tags "MifareTag *tags"
.\"  ____                      _       _   _
.\" |  _ \  ___  ___  ___ _ __(_)_ __ | |_(_) ___  _ __
.\" | | | |/ _ \/ __|/ __| '__| | '_ \| __| |/ _ \| '_ \
.\" | |_| |  __/\__ \ (__| |  | | |_) | |_| | (_) | | | |
.\" |____/ \___||___/\___|_|  |_| .__/ \__|_|\___/|_| |_|
.\"                             |_|
.Sh DESCRIPTION
The
.Fn freefare_*
family of functions allow agnostic access to the
.Vt MifareTag
present on a given NFC device.
.Pp
.Fn freefare_get_tags
returns a NULL-terminated list of
.Vt MifareTag
present on
.Vt device .
This list has to be freed after usage by either:
.Bl -hyphen
.It
calling the
.Fn freefare_free_tags
function.  All tags in the list are automatically freed;
.It
calling the
.Fn freefare_free_tag
function to free each tag in the list individually, and freeing the list itself
using the
.Fn free
function.
.El
.Pp
Because of the nature of the target detection process, any previously detected
target will be in an inconsistent state after a call to
.Fn freefare_get_tags .
It is the programmer's responsibility to free these targets before calling the
.Fn freefare_get_tags
function.
.Pp
Information about a given
.Vt MifareTag
can be gathered using the
.Fn freefare_get_tag_type ,
.Fn freefare_get_tag_uid
and
.Fn freefare_get_tag_friendly_name
functions.
.\"  ____      _                                 _
.\" |  _ \ ___| |_ _   _ _ __ _ __   __   ____ _| |_   _  ___  ___
.\" | |_) / _ \ __| | | | '__| '_ \  \ \ / / _` | | | | |/ _ \/ __|
.\" |  _ <  __/ |_| |_| | |  | | | |  \ V / (_| | | |_| |  __/\__ \
.\" |_| \_\___|\__|\__,_|_|  |_| |_|   \_/ \__,_|_|\__,_|\___||___/
.\"
.Sh RETURN VALUES
Unless stated otherwise, all functions return a value greater than or equal to
.Va 0
on success or
.Va -1
on failure.
.\"  ____                    _
.\" / ___|  ___  ___    __ _| |___  ___
.\" \___ \ / _ \/ _ \  / _` | / __|/ _ \
.\"  ___) |  __/  __/ | (_| | \__ \ (_) |
.\" |____/ \___|\___|  \__,_|_|___/\___/
.\"
.Sh SEE ALSO
.Xr free 3 ,
.Xr mifare_classic 3 ,
.Xr mifare_ultralight 3
.\"     _         _   _
.\"    / \  _   _| |_| |__   ___  _ __ ___
.\"   / _ \| | | | __| '_ \ / _ \| '__/ __|
.\"  / ___ \ |_| | |_| | | | (_) | |  \__ \
.\" /_/   \_\__,_|\__|_| |_|\___/|_|  |___/
.\"
.Sh AUTHORS
.An Romain Tartiere Aq romain@blogreen.org
.An Romuald Conty Aq rconty@il4p.fr


NEWS
Changes between 0.3.4 and 0.3.5  [20 jan 2013]

  *) Depends on libnfc 1.7.0

Changes between 0.3.3 and 0.3.4  [1 mar 2012]

  *) Fix Mifare Ultralight C card detection routine

Changes between 0.3.2 and 0.3.3  [1 mar 2012]

  *) Depends on libnfc 1.6.0

Changes between 0.3.1 and 0.3.2  [17 oct 2011]

  *) New API functions to ease up creation of applications with strong
     cryptography on Mifare DESFire EV1:
     mifare_desfire_create_application_3k3des(),
     mifare_desfire_create_application_aes().
  *) New API functions for ISO 7616 support om Mifare DESFire EV1:
     mifare_desfire_create_application_iso(),
     mifare_desfire_create_application_3k3des_iso(),
     mifare_desfire_create_application_aes_iso(),
     mifare_desfire_get_df_names(),
     mifare_desfire_get_iso_file_ids(),
     mifare_desfire_create_std_data_file_iso(),
     mifare_desfire_create_backup_data_file_iso(),
     mifare_desfire_create_linear_record_file_iso(),
     mifare_desfire_create_cyclic_record_file_iso().
  *) New public API function freefare_tag_new();
  *) Microsoft Windows support.

Changes between 0.3.0 and 0.3.1  [23 feb 2011]

  *) Fix mifare_classic_transfer() for devices returning a 1 byte response on
     success.
  *) New API function mifare_desfire_last_pcd_error().

Changes between 0.2.3 and 0.3.0  [23 dec 2010]

  *) Add support for ISO and AES authentication through
     mifare_desfire_authenticate_iso() and mifare_desfire_authenticate_aes().
  *) Add support for 3K3DES and AES cryptographic operations and add new
     functions mifare_desfire_3k3des_key_new(),
     mifare_desfire_3k3des_key_new_with_version(), mifare_desfire_aes_key_new()
     and mifare_desfire_aes_key_new_with_version() for generating keys.
  *) New functions mifare_desfire_free_mem(),
     mifare_desfire_set_configuration(), mifare_desfire_set_default_key(),
     mifare_desfire_set_ats(), mifare_desfire_get_card_uid(), for Mifare
     DESFire EV1 targets manipulation.
  *) Deprecate authentication information when deleting the currently selected
     application

Changes between 0.2.2 and 0.2.3  [23 dec 2010]

  *) Fix a buffer overflow in mifare_desfire_get_file_settings().

Changes between 0.2.1 and 0.2.2  [23 nov 2010]

  *) Fix build when cutter is not available or is disabled.

Changes between 0.2.0 and 0.2.1  [23 nov 2010]

  *) The mifare_desfire_error_lookup() and mifare_desfire_get_last_error()
     functions were removed and replaced by the freefare_strerror(),
     freefare_strerror_r() and freefare_perror() functions.
  *) The library reports errors other that the ones returned by the PICC.
  *) The MDAD_KEYx macro where renamed MDAR_KEYx for consistency.
  *) The MDCM_MACING macro was renamed MDCM_MACED.
  *) The MDCM_FULLDES macro was renamed MDCM_ENCIPHERED.
  *) New function mifare_desfire_last_picc_error().
  *) New function mifare_desfire_aid_get_aid().

Changes between 0.1.0 and 0.2.0  [07 sep 2010]

  *) New function freefare_free_tag() to free individual tags from a MifareTag
     list.
  *) The mifare_application_alloc() size parameter is now expressed in bytes
     and not in sectors.
  *) New API function mad_sector_reserved().
  *) The mifare_classic_format_sector() functions now require a sector number
     instead of a block number.
  *) New API functions mifare_classic_block_sector(),
     mifare_classic_sector_first_block(), mifare_classic_sector_block_count()
     and mifare_classic_sector_last_block().
  *) New API functions mifare_application_read(), mifare_application_write().


libfreefare/freefare.c
/*-
 * Copyright (C) 2010, Romain Tartiere, Romuald Conty.
 *
 * This program is free software: you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as published by the
 * Free Software Foundation, either version 3 of the License, or (at your
 * option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
 * more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>
 *
 * $Id$
 */

#include <stdlib.h>
#include <string.h>

#include "config.h"

#ifdef HAVE_LIBNFC
#include <nfc/nfc.h>
#include <freefare.h>
#endif
#ifdef HAVE_PCSC
#include "freefare_pcsc.h"
#endif

#include <reader.h>

#include "freefare_internal.h"

#ifdef HAVE_PCSC
#include "freefare_pcsc_tags.h"
#endif

#define MAX_CANDIDATES 16

#define NXP_MANUFACTURER_CODE 0x04

struct supported_tag supported_tags[] = {
    { CLASSIC_1K,   "Mifare Classic 1k",            0x08, 0, 0, { 0x00 }, NULL },
    { CLASSIC_1K,   "Infineon Mifare Classic 1k",   0x88, 0, 0, { 0x00 }, NULL },
    { CLASSIC_4K,   "Mifare Classic 4k",            0x18, 0, 0, { 0x00 }, NULL },
    { CLASSIC_4K,   "Mifare Classic 4k (Emulated)", 0x38, 0, 0, { 0x00 }, NULL },
    { DESFIRE,      "Mifare DESFire",               0x20, 5, 4, { 0x75, 0x77, 0x81, 0x02 /*, 0xXX */ }, NULL},
#if (!defined(is_mifare_ultralightc_on_reader))
    { ULTRALIGHT_C, "Mifare UltraLightC",           0x00, 0, 0, { 0x00 }, NULL },
#else
    { ULTRALIGHT_C, "Mifare UltraLightC",           0x00, 0, 0, { 0x00 }, is_mifare_ultralightc_on_reader },
#endif
    { ULTRALIGHT,   "Mifare UltraLight",            0x00, 0, 0, { 0x00 }, NULL },
};

#ifdef HAVE_LIBNFC
/*
 * Automagically allocate a MifareTag given a device and target info.
 */
MifareTag
freefare_tag_new (nfc_device *device, nfc_iso14443a_info nai)
{
    bool found = false;
    struct supported_tag *tag_info;
    MifareTag tag;

    /* Ensure the target is supported */
    for (size_t i = 0; i < sizeof (supported_tags) / sizeof (struct supported_tag); i++) {
	if (((nai.szUidLen == 4) || (nai.abtUid[0] == NXP_MANUFACTURER_CODE)) &&
	    (nai.btSak == supported_tags[i].SAK) &&
	    (!supported_tags[i].ATS_min_length || ((nai.szAtsLen >= supported_tags[i].ATS_min_length) &&
						   (0 == memcmp (nai.abtAts, supported_tags[i].ATS, supported_tags[i].ATS_compare_length)))) &&
	    ((supported_tags[i].check_tag_on_reader == NULL) ||
	     supported_tags[i].check_tag_on_reader(device, nai))) {

	    tag_info = &(supported_tags[i]);
	    found = true;
	    break;
	}
    }

    if (!found)
	return NULL;

    /* Allocate memory for the found MIFARE target */
    switch (tag_info->type) {
    case CLASSIC_1K:
    case CLASSIC_4K:
	tag = mifare_classic_tag_new ();
	break;
    case DESFIRE:
	tag = mifare_desfire_tag_new ();
	break;
    case ULTRALIGHT:
    case ULTRALIGHT_C:
	tag = mifare_ultralight_tag_new ();
	break;
    }

    if (!tag)
	return NULL;

    /*
     * Initialize common fields
     * (Target specific fields are initialized in mifare_*_tag_new())
     */
    tag->device = device;
    tag->info = nai;
    tag->active = 0;
    tag->tag_info = tag_info;

    return tag;
}
#endif
#ifdef HAVE_PCSC
MifareTag
freefare_tag_new_pcsc (struct pcsc_context *context, const char *reader)
{
    struct supported_tag *tag_info = NULL;
    enum mifare_tag_type tagtype;
    bool found = false;
    uint8_t buf[] = { 0xFF, 0xCA, 0x00, 0x00, 0x00 };
    uint8_t ret[12];
    unsigned char pbAttr[MAX_ATR_SIZE];
    unsigned int k;
    char crc = 0x00;
    size_t size;
    MifareTag tag;
    LONG err;
    DWORD atrlen = sizeof(pbAttr);
    DWORD dwActiveProtocol;
    DWORD retlen;
    SCARDHANDLE hCard;
    SCARD_IO_REQUEST sendpci;
    SCARD_IO_REQUEST ioreq;

    err = SCardConnect(context->context, reader, SCARD_SHARE_SHARED, 
			SCARD_PROTOCOL_T0 | SCARD_PROTOCOL_T1, &hCard, &dwActiveProtocol);
    if(err)
	return NULL;

    switch (dwActiveProtocol)
    {
	case SCARD_PROTOCOL_T0: sendpci = *SCARD_PCI_T0;
	     break;
	case SCARD_PROTOCOL_T1: sendpci = *SCARD_PCI_T1;
	     break;
    }

    /* get and card uid */
    retlen = sizeof(ret);
    err = SCardTransmit(hCard, &sendpci, buf, sizeof(buf), NULL /*&ioreq*/, ret, &retlen);
    if (err)
    {
	return NULL;
    }

    err = SCardGetAttrib ( hCard , SCARD_ATTR_ATR_STRING, (unsigned char *) &pbAttr, &atrlen );
    if (err)
    {
	return NULL;
    }

    found = false;
    for (k = 0; pcsc_supported_atrs[k].len != 0; k++){
	if (atrlen != pcsc_supported_atrs[k].len) { 
	    continue;
	}
	if ( pcsc_supported_atrs[k].mask == NULL ){
	    /* no bitmask here */
	    if ( ! memcmp(pcsc_supported_atrs[k].tag ,pbAttr ,atrlen) ) {
		tagtype = pcsc_supported_atrs[k].type;
		found = true;
		break;
	    }
	} else {
	    /* bitmask case */
	    int c = 0;
	    for (c = 0; c < pcsc_supported_atrs[k].len; c++){
		if((pcsc_supported_atrs[k].tag[c] & pcsc_supported_atrs[k].mask[c]) != (pbAttr[c] & pcsc_supported_atrs[k].mask[c])){
		    break;
		}
	    }
	    if (c == pcsc_supported_atrs[k].len) {
		tagtype = pcsc_supported_atrs[k].type;
		found = true;
		break;
	    }
	}
    }	
    if (!found) { 
	return NULL;
    }

    found = false;
    for (size = 0; size < sizeof (supported_tags) / sizeof (struct supported_tag); size++) {
	if(supported_tags[size].type == tagtype) {
	    tag_info = &(supported_tags[size]);
	    found = true;
	    break;
	}
    }

    if(!found)
	return NULL;

    for (k = 1 /*! 1. Byte wird ignoriert*/ ; k < atrlen; k++ )
    {
	crc ^= pbAttr[k];
    }
    if (crc)
    	return NULL;

    /* Allocate memory for the found MIFARE target */
    switch (tag_info->type) {
    case CLASSIC_1K:
    case CLASSIC_4K:
	return NULL;
	/* classic tags not yet supported with PCSC */
	/* tag = mifare_classic_tag_new (); */
	break;
    case DESFIRE:
	tag = mifare_desfire_tag_new ();
	break;
    case ULTRALIGHT:
    case ULTRALIGHT_C:
	return NULL;
	/* ultralight tags not yet supported with PCSC */
	/* tag = mifare_ultralight_tag_new (); */
	break;
    }

    if (!tag)
	return NULL;

    /*
     * Initialize common fields
     * (Target specific fields are initialized in mifare_*_tag_new())
     */
#ifdef HAVE_LIBNFC
    memcpy(tag->info.abtUid, ret, retlen - 2);
    tag->info.szUidLen = retlen - 2;
    tag->device = NULL;
#endif
    tag->hContext = context->context;
    tag->hCard = hCard;
    tag->active = 0;
    tag->tag_info = tag_info;
    FILL_SZREADER(tag, reader);

    tag->lastPCSCerror = SCardDisconnect(tag->hCard, SCARD_LEAVE_CARD);

    return tag;
}
#endif
/*
 * MIFARE card common functions
 *
 * The following functions send NFC commands to the initiator to prepare
 * communication with a MIFARE card, and perform required cleannups after using
 * the targets.
 */
#ifdef HAVE_LIBNFC
/*
 * Get a list of the MIFARE targets near to the provided NFC initiator.
 *
 * The list has to be freed using the freefare_free_tags() function.
 */
MifareTag *
freefare_get_tags (nfc_device *device)
{
    MifareTag *tags = NULL;
    int tag_count = 0;

    nfc_initiator_init(device);

    // Drop the field for a while
    nfc_device_set_property_bool(device,NP_ACTIVATE_FIELD,false);

    // Configure the CRC and Parity settings
    nfc_device_set_property_bool(device,NP_HANDLE_CRC,true);
    nfc_device_set_property_bool(device,NP_HANDLE_PARITY,true);
    nfc_device_set_property_bool(device,NP_AUTO_ISO14443_4,true);

    // Enable field so more power consuming cards can power themselves up
    nfc_device_set_property_bool(device,NP_ACTIVATE_FIELD,true);

    // Poll for a ISO14443A (MIFARE) tag
    nfc_target candidates[MAX_CANDIDATES];
    int candidates_count;
    nfc_modulation modulation = {
	.nmt = NMT_ISO14443A,
	.nbr = NBR_106
    };
    if ((candidates_count = nfc_initiator_list_passive_targets(device, modulation, candidates, MAX_CANDIDATES)) < 0)
	return NULL;

    tags = malloc(sizeof (void *));
    if(!tags) return NULL;
    tags[0] = NULL;

    for (int c = 0; c < candidates_count; c++) {
	MifareTag t;
	if ((t = freefare_tag_new(device, candidates[c].nti.nai))) {
	    /* (Re)Allocate memory for the found MIFARE targets array */
	    MifareTag *p = realloc (tags, (tag_count + 2) * sizeof (MifareTag));
	    if (p)
		tags = p;
	    else
		return tags; // FAIL! Return what has been found so far.
	    tags[tag_count++] = t;
	    tags[tag_count] = NULL;
	}
    }

    return tags;
}
#endif
#ifdef HAVE_PCSC
/*
 * Get a list of the MIFARE targets near to the provided NFC initiator.
 * (Usally its just one tag, because pcsc can not detect more)
 * phContext must be established with SCardEstablishContext before 
 * calling this function.
 * mszReader is the Name of the SmartCard Reader to use
 * The list has to be freed using the freefare_free_tags() function.
 */
MifareTag *
freefare_get_tags_pcsc (struct pcsc_context *context, const char *reader)
{
    MifareTag *tags = NULL;
    
    tags = (MifareTag *)malloc(2*sizeof (MifareTag));
    if(!tags)
    {
	return NULL;
    }
    tags[0] = freefare_tag_new_pcsc(context, reader);
    tags[1] = NULL;
    if(tags[0] == NULL)
    	return NULL;

    return tags;
}
#endif
/*
 * Returns the type of the provided tag.
 */
enum mifare_tag_type
freefare_get_tag_type (MifareTag tag)
{
    return tag->tag_info->type;
}

/*
 * Returns the friendly name of the provided tag.
 */
const char *
freefare_get_tag_friendly_name (MifareTag tag)
{
    return tag->tag_info->friendly_name;
}

/*
 * Returns the UID of the provided tag.
 */
char *
freefare_get_tag_uid (MifareTag tag)
{
    size_t i;
   	char *res = (char *)malloc (2 * tag->info.szUidLen + 1);
   	for (i =0; i < tag->info.szUidLen; i++)
       	snprintf (res + 2*i, 3, "%02x", tag->info.abtUid[i]);
   	return res;
}

/*
 * Free the provided tag.
 */
void
freefare_free_tag (MifareTag tag)
{
    if (tag) {
        switch (tag->tag_info->type) 
	{
	    case CLASSIC_1K:
	    case CLASSIC_4K:
		mifare_classic_tag_free (tag);
	    	break;
	    case DESFIRE:
		mifare_desfire_tag_free (tag);
	    	break;
	    case ULTRALIGHT:
	    case ULTRALIGHT_C:
		mifare_ultralight_tag_free (tag);
		break;
        }
	FREE_SZREADER(tag->szReader);
    }
}

const char *
freefare_strerror (MifareTag tag)
{
    const char *p = "Unknown error";
#if defined(HAVE_LIBNFC) && defined(HAVE_PCSC)
    if(tag->device != NULL) // we use libnfc
#endif
#ifdef HAVE_LIBNFC
    {
	if (nfc_device_get_last_error (tag->device) < 0) {
	    p = nfc_strerror (tag->device);
	} else {
	    if (tag->tag_info->type == DESFIRE) {
	    	if (MIFARE_DESFIRE (tag)->last_pcd_error) {
		    p = mifare_desfire_error_lookup (MIFARE_DESFIRE (tag)->last_pcd_error);
	    	} else if (MIFARE_DESFIRE (tag)->last_picc_error) {
	            p = mifare_desfire_error_lookup (MIFARE_DESFIRE (tag)->last_picc_error);
	    	}
	    }
	}
    }
#endif
#if defined(HAVE_LIBNFC) && defined(HAVE_PCSC)
    else // we use the pcsc protocol
#endif
#ifdef HAVE_PCSC
    {
	if (tag->lastPCSCerror != 0){
#ifdef _WIN32
    return "Internal PCSC error";
#else
	    p = (const char*) pcsc_stringify_error(tag->lastPCSCerror);
    	    return p;
#endif
	} else {
	     if (tag->tag_info->type == DESFIRE) {
		if (MIFARE_DESFIRE (tag)->last_pcd_error) {
		    p = mifare_desfire_error_lookup (MIFARE_DESFIRE (tag)->last_pcd_error);
		} else if (MIFARE_DESFIRE (tag)->last_picc_error) {
		    p = mifare_desfire_error_lookup (MIFARE_DESFIRE (tag)->last_picc_error);
		}   
	    }   
	}
    }
#endif
    return p;
}

int
freefare_strerror_r (MifareTag tag, char *buffer, size_t len)
{
    return (snprintf (buffer, len, "%s", freefare_strerror (tag)) < 0) ? -1 : 0;
}

void
freefare_perror (MifareTag tag, const char *string)
{
    fprintf (stderr, "%s: %s\n", string, freefare_strerror (tag));
}

/*
 * Free the provided tag list.
 */
void
freefare_free_tags (MifareTag *tags)
{
  int i;
    if (tags) {
	for (i=0; tags[i]; i++) {
	    freefare_free_tag(tags[i]);
	}
	free (tags);
    }
}

/*
 * create context for pcsc readers
 */
#ifdef HAVE_PCSC
void
pcsc_init(struct pcsc_context** context)
{
	LONG err;
	struct pcsc_context *con =  (struct pcsc_context *)malloc(sizeof(struct pcsc_context));
	err = SCardEstablishContext(SCARD_SCOPE_SYSTEM, NULL, NULL, &con->context);
	if (err < 0)
	{
		*context = NULL;
		return;
	}
	*context = con;
}

/*
 * destroy context for pcsc readers
 */
void
pcsc_exit(struct pcsc_context* context)
{
	if (context->readers)
    #ifndef __APPLE__
		SCardFreeMemory(context->context, context->readers);
    #endif
	SCardReleaseContext(context->context);
}

/*
 * list pcsc devices
 */

LONG
pcsc_list_devices(struct pcsc_context* context, LPSTR *string)
{
    LONG err;
    LPSTR str = NULL;
    DWORD size;
    static char empty[] = "\0";

    err = SCardListReaders(context->context, NULL, NULL, &size);
    str = (char *)malloc(sizeof(char) * size);
    if (!str)
    {
    	context->readers = NULL;
	    *string = empty;
    	return SCARD_E_NO_MEMORY;
    }
    err = SCardListReaders(context->context, NULL, str, &size);
    if (err != SCARD_S_SUCCESS)
    {
      context->readers = NULL;
      *string = empty;
    }
    else
    {
      *string = str;
      context->readers = str;
    }
    return err;
}
#endif
/*
 * Low-level API
 */

void *
memdup (const void *p, const size_t n)
{
    void *res;
    if ((res = malloc (n))) {
	memcpy (res, p, n);
    }
    return res;
}
